bin/setupReplication.js

```diff
@@ -0,0 +1,449 @@
+                }
+            ),
+            (srcArn, next) => this._isValidRole('source', srcArn,
+                (err, tgtArn) => {
+                    next(err, tgtArn);
+                }
+            ),
+            (tgtArn, next) => this._isValidRole('target', tgtArn,
+                (err, res) => {
+                    next(err, res);
+                }
+            ),
+        ], (err, res) => {
+            cb(err, res);
+        });
+    }
+
+    _isValidBucket(where, cb) {
+        // Does the bucket exist and is it reachable?
+        const bucket = where === 'source' ? this._sourceBucket :
+            this._targetBucket;
+        this._s3[where].headBucket({ Bucket: bucket }, err => {
+            if (err) {
+                this._log.error('bucket sanity check error', {
+                    method: '_SetupReplication._isValidBucket',
+                    error: err.message,
+                    errStack: err.stack,
+                });
+                return cb(err);
+            }
+            return cb();
+        });
+    }
+
+    _isVersioningEnabled(where, cb) {
+        // Does the bucket have versioning enabled?
+        const bucket = where === 'source' ? this._sourceBucket :
+            this._targetBucket;
+        this._s3[where].getBucketVersioning({ Bucket: bucket }, (err, res) => {
+            if (err || res.Status === 'Disabled') {
+                this._log.error('versioning sanity check error', {
+                    method: '_SetupReplication._isVersioningEnabled',
+                    error: err.message,
+                    errStack: err.stack,
+                });
+                return cb(err);
+            }
+            return cb();
+        });
+    }
+
+    _isValidRole(where, arn, cb) {
+        // Is the role mentioned in the replication config available in IAM
+
+        // Goal is to get Role given known ARN.
+        // If err, there is no matching role
+        let roleName;
+        let tgtArn;
+
+        if (where === 'source') {
+            [roleName, tgtArn] = arn.split(',');
+        } else if (where === 'target') {
+            roleName = arn;
+        }
+        const namesOnly = roleName.split('/').pop();
+
+        this._iam[where].getRole({ RoleName: namesOnly }, (err, res) => {
+            if (err || roleName !== res.Role.Arn) {
+                this._log.error('role validation sanity check error', {
+                    method: '_SetupReplication._isValidRole',
+                    error: err.message,
+                    errStack: err.stack,
+                });
+                return cb(err);
+            }
+            return cb(err, tgtArn);
+        });
+    }
+
+    _isReplicationEnabled(src, cb) {
+        // Is the Replication config enabled?
+        this._s3[src].getBucketReplication({ Bucket: this._sourceBucket },
+            (err, res) => {
+                const r = res.ReplicationConfiguration;
+                if (err || r.Rules[0].Status === 'Disabled') {
```

Comment:
Ya I agree, especially since the params for replication are nearly hard-coded in. I wasn't sure if there would be scenarios where this might ever be 'Disabled'. If it's cool, I'll remove it and only check on `err`?